core/src/apps/zcash/unified_addresses.py

```diff
@@ -0,0 +1,137 @@
+"""
+Implementation of encoding and decoding of Zcash
+unified addresses according to the ZIP-316.
+see: https://zips.z.cash/zip-0316
+"""
+
+from typing import TYPE_CHECKING
+
+from trezor.crypto.bech32 import Encoding, bech32_decode, bech32_encode, convertbits
+from trezor.utils import BufferReader, empty_bytearray
+from trezor.wire import DataError
+
+from apps.common.coininfo import CoinInfo
+from apps.common.readers import read_compact_size
+from apps.common.writers import write_bytes_fixed, write_compact_size
+
+from .f4jumble import f4jumble, f4unjumble
+
+if TYPE_CHECKING:
+    from enum import IntEnum
+else:
+    IntEnum = object
+
+
+class Typecode(IntEnum):
+    P2PKH = 0x00
+    P2SH = 0x01
+    SAPLING = 0x02
+    ORCHARD = 0x03
+
+
+def receiver_length(typecode: int) -> int | None:
+    """Byte length of a receiver."""
+    if typecode in (Typecode.P2PKH, Typecode.P2SH):
+        return 20
+    if typecode in (Typecode.SAPLING, Typecode.ORCHARD):
+        return 43
+    return None
+
+
+def prefix(coin: CoinInfo) -> str:
+    """Prefix for a unified address."""
+    if coin.coin_name == "Zcash":
+        return "u"
+    if coin.coin_name == "Zcash Testnet":
+        return "utest"
+    raise ValueError
+
+
+def padding(hrp: str) -> bytes:
+    assert len(hrp) <= 16
+    return hrp.encode() + bytes(16 - len(hrp))
+
+
+def encode(receivers: dict[Typecode, bytes], coin: CoinInfo) -> str:
+    # multiple transparent receivers forbidden
+    assert not (Typecode.P2PKH in receivers and Typecode.P2SH in receivers)
+    # at least one shielded address must be present
+    assert Typecode.SAPLING in receivers or Typecode.ORCHARD in receivers
+
+    length = 16  # 16 bytes for padding
+    for typecode, receiver_bytes in receivers.items():
+        length += 2  # typecode (1 byte) + length (1 byte)
+        length += receiver_length(typecode) or len(receiver_bytes)
```

Comment:
nit: I'd simplify this for better legibility, since it's only to estimate the size for buffer preallocation.
```suggestion
    length = 16  # 16 bytes for padding
    for receiver_bytes in receivers.values():
        length += 2  # typecode (1 byte) + length (1 byte)
        length += len(receiver_bytes)
```

Reply:
[b70a2c1](https://github.com/trezor/trezor-firmware/pull/2398/commits/b70a2c15d742251515ceb49d4b3a96a88cdf2d16)